Add DetailedGuide presenter.

tvararu opened the pull request:
This is implemented very similarly to the `CaseStudy` presenter, which inherits from the same base `Edition` presenter.

Some of the methods (`first_public_at`, `body`) can potentially be refactored into the parent presenter.

Depends on a content-schema update, and will fail until it's merged. (https://github.com/alphagov/govuk-content-schemas/pull/279)

Relevant Trello ticket: https://trello.com/c/yAU5zVvg/315-6-detailed-guides-migration-implement-publishing-of-format-to-publishing-api-large

boffbowsh commented on app/presenters/publishing_api_presenters/detailed_guide.rb:
```diff
@@ -0,0 +1,36 @@
+  end
+
+  def details
+    super.merge({
```
The braces aren't required here

boffbowsh commented on app/presenters/publishing_api_presenters/detailed_guide.rb:
```diff
@@ -0,0 +1,36 @@
+class PublishingApiPresenters::DetailedGuide < PublishingApiPresenters::Edition
+
+  def links
+    extract_links([
```
This could be one line

boffbowsh commented:
Still needs a lot of links changes as per the schema, like related guides, related mainstream content etc. It's also failing the linter at present

tvararu commented:
Related guides have been fixup'd into the main presenter commit.

Related mainstream content has been added in the latest commit, https://github.com/alphagov/whitehall/pull/2553/commits/60b94b87d6a3084575a5080d17b975460a05ba02, and will be fixup'd if this implementation aligns with the content schema changes described in https://github.com/alphagov/govuk-content-schemas/pull/279#issuecomment-211971848.

tvararu commented:
I have removed the earlier `related_mainstream_content` refactoring and I've replaced it with an implementation that properly queries the `content-store` to fetch their IDs, hopefully. Check https://github.com/alphagov/whitehall/pull/2553/commits/1521d06b98d2655659fd0fe8b7c8bfa542c13a89#diff-3e81f395ae3d0ab058db8e5adfe8f55dR53.

It needs a bit more refactoring, such as moving the URL parsing inside the model to DRY it up and test it with more complex URLs.

tijmenb commented on app/presenters/publishing_api_presenters/links_presenter.rb:
```diff
@@ -42,6 +46,33 @@ def policy_area_ids
       (item.try(:topics) || []).map(&:content_id)
     end
 
+    def related_guides_ids
```
Do you know if there are more formats using related guides & friends? If not it might be better to move this to `PublishingApiPresenters::DetailedGuide`.

tijmenb commented on app/presenters/publishing_api_presenters/links_presenter.rb:
```diff
@@ -42,6 +46,33 @@ def policy_area_ids
+      item.try(:related_detailed_guide_ids) || []
+    end
+
+    def related_mainstream_ids
```
There's an endpoint on publishing-api to do this more efficiently:

http://www.rubydoc.info/gems/gds-api-adapters/GdsApi/PublishingApiV2#lookup_content_ids-instance_method

As a rule, publishing-apps should only talk to publishing-api and frontend apps should only talk to content-store.

tvararu commented:
Updated:

- `change_note` removed from the `Edition` presenter and placed only in the `CaseStudy` presenter, to meet https://github.com/alphagov/govuk-content-schemas/pull/279#discussion-diff-59350994.
- Related mainstream content link IDs are now generated using a `PublishingApiV2` lookup instead of a `ContentStore` one.
- `related_guides` and `related_mainstream` no longer in the catch-all `links_presenter`, but in the `DetailedGuide` presenter.
- `related_mainstream_content` now correctly generated, to meet https://github.com/alphagov/govuk-content-schemas/pull/279#discussion_r59350751. The contents of the array should be sorted to the best of my knowledge; we query `PublishingApiV2` with an ordered array of paths (https://github.com/alphagov/whitehall/pull/2553/files#diff-da11b3576e8a1d61c49f7457c686b02eR51), and it returns a hash which appear to be deterministically in the same order.

boffbowsh commented on app/models/detailed_guide.rb:
```diff
@@ -76,8 +76,26 @@ def translatable?
+  def parse_base_path_from_related_mainstream_url(url)
+    return nil if url.nil? || url.empty?
+    parsed_url = URI.parse(url)
+    url_is_invalid = !parsed_url.host.include?('gov.uk')
```
This should only be things that are hosted on GOV.UK itself. This would consider URLs on e.g. https://www.ofgem.gov.uk/ valid.

tvararu replied:
So `gov.uk` and `www.gov.uk` only then.

boffbowsh replied:
Yep

boffbowsh commented on app/presenters/publishing_api_presenters/detailed_guide.rb:
```diff
@@ -0,0 +1,56 @@
+    base_paths.push(item.related_mainstream_base_path)
+    base_paths.push(item.additional_related_mainstream_base_path)
+
+    if base_paths.any? # NB: Is false for [nil, nil].
```
Should add a `base_paths.compact!` before this to remove the nils.

boffbowsh commented on app/presenters/publishing_api_presenters/detailed_guide.rb:
```diff
@@ -0,0 +1,56 @@
+
+    if base_paths.any? # NB: Is false for [nil, nil].
+      publishing_api = Whitehall.publishing_api_v2_client
+      publishing_api.lookup_content_ids(base_paths: base_paths).values
```
Also probably want `values.compact` here for things that weren't in publishing-api

boffbowsh commented on app/presenters/publishing_api_presenters/detailed_guide.rb:
```diff
@@ -0,0 +1,56 @@
+  end
+
+  def related_guides
+    item.try(:related_detailed_guide_ids) || []
```
These will be integer ids in Whitehall I think, not content ids.

boffbowsh replied:
Related guides needs a test case later.

tvararu replied:
Let's fix+test and add the behaviour separately, in the interest of merging this ASAP.

I'm going to change it to return just `[]` so it's clear that it's unimplemented.

boffbowsh commented on app/presenters/publishing_api_presenters/case_study.rb:
```diff
@@ -27,6 +27,7 @@ def details
       format_display_type: item.display_type_key,
       first_public_at: first_public_at,
       change_history: item.change_history.as_json,
+      change_note: item.most_recent_change_note,
```
Looking at the govt frontend code, I don't believe this is actually used either, and it's not required in the schemas. I think we should remove it here and remove it from the schemas as another PR at some point.

fofr replied:
I assume the most recent item in the change_history would be the same as the change_note?

tvararu replied:
Yes.

boffbowsh commented:
Looks ok apart from the `# TODO`

tvararu commented:
Done. ✅

boffbowsh commented:
Tests are failing

tvararu commented:
`change_note` removal stuff in `PublishingApiPresenters::DocumentCollectionPlaceholderTest`, even though it passed earlier. That's weird.

tvararu commented:
💚

boffbowsh commented:
👍

fofr commented on app/presenters/publishing_api_presenters/detailed_guide.rb:
```diff
@@ -0,0 +1,71 @@
+  end
+
+  def body
+    Whitehall::GovspeakRenderer.new.govspeak_edition_to_html(item)
```
I think the `body` presenter needs to use `govspeak_with_attachments_to_html` so that embedded attachments render correctly.

fofr replied:
Hmm, looks like `bare_govspeak_edition_to_html` calls `edition_body_with_attachments_and_alt_format_information` so might be fine.
